infra/cifuzz/filestore/github_actions/__init__.py

```diff
@@ -0,0 +1,84 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Implementation of a filestore using Github actions artifacts."""
+import os
+import logging
+
+import http_utils
+import filestore
+from filestore.github_actions import github_api
+from third_party.github_actions_toolkit.artifact import artifact_client
+
+
+class GithubActionsFilestore(filestore.BaseFilestore):
+  """Implementation of BaseFilestore using Github actions artifacts. Relies on
+  github_actions_toolkit for using the GitHub actions API and the github_api
+  module for using GitHub's standard API. We need to use both because the GitHub
+  actions API is the only way to upload an artifact but it does not support
+  downloading artifacts from other runs. The standard GitHub API does support
+  this however."""
+
+  def __init__(self, config):
+    super().__init__(config)
+    self.github_api_http_headers = github_api.get_http_auth_headers(config)
+
+  def upload_directory(self, name, directory):  # pylint: disable=no-self-use
+    """Uploads |directory| as artifact with |name|."""
+    directory = os.path.abspath(directory)
+
+    # Get file paths.
+    file_paths = []
+    for root, _, curr_file_paths in os.walk(directory):
+      for file_path in curr_file_paths:
+        file_paths.append(os.path.join(root, file_path))
+
+    logging.debug('file_paths: %s', file_paths)
+
+    # TODO(metzman): Zip so that we can upload directories within directories
+    # and save time?
+
+    return artifact_client.upload_artifact(name, file_paths, directory)
+
+  def download_corpus(self, name, dst_directory):  # pylint: disable=unused-argument,no-self-use
+    """Downloads the corpus located at |name| to |dst_directory|."""
+    logging.debug('listing artifact')
+    artifacts = self._list_artifacts()
+    corpus_artifact = github_api.find_artifact(name, artifacts)
+    logging.debug('Corpus artifact: %s.', corpus_artifact)
+    if not corpus_artifact:
+      logging.warning('Could not download corpus: %s.', name)
+      return False
+    url = corpus_artifact['archive_download_url']
+    logging.debug('Corpus artifact url: %s.', url)
+    return http_utils.download_and_unpack_zip(
+        url, dst_directory, headers=self.github_api_http_headers)
+
+  def _list_artifacts(self):
+    """Returns a list of artifacts."""
+    return github_api.list_artifacts(self.config.project_repo_owner,
+                                     self.config.project_repo_name,
+                                     self.github_api_http_headers)
+
+  def download_latest_build(self, name, dst_directory):
+    """Downloads latest build with name |name| to |dst_directory|."""
+    artifacts = self._list_artifacts()
+    build_artifact = github_api.find_artifact(name, artifacts)
```

Comment:
add a convenience function to github_api to do both the list_artifacts and find_artifact steps since it's done twice in this file?

Reply:
Done. I ended up making both download_corpus and download_latest_build wrappers around download_artifact.